File: Java/Problem_052.java
import java.util.Arrays;

public class Euler52 {

    static int[] signature(int x) {
        int[] cnt = new int[10];
        while (x > 0) {
            cnt[x % 10]++;
            x /= 10;
        }
        return cnt;
    }

    static boolean sameSig(int[] a, int[] b) {
        return Arrays.equals(a, b);
    }

    static int solve() {
        for (int n = 1; ; n++) {
            int low = 1;
            for (int i = 1; i < n; i++) low *= 10;

            int p10 = 1;
            for (int i = 0; i < n; i++) p10 *= 10;

            int high = (p10 - 1) / 6;

            for (int x = low; x <= high; x++) {
                int[] sig = signature(x);
                boolean ok = true;

                for (int k = 2; k <= 6; k++) {
                    if (!sameSig(signature(k * x), sig)) {
                        ok = false;
                        break;
                    }
                }
                if (ok) return x;
            }
        }
    }

    public static void main(String[] args) {
        System.out.println(solve()); // 142857
    }
}

File: Python/Problem_052.py
def signature(num: int) -> tuple[int, ...]:
    cnt = [0] * 10
    while num > 0:
        cnt[num % 10] += 1
        num //= 10
    return tuple(cnt)

def euler_52() -> int:
    n = 1
    while True:
        low = 10 ** (n - 1)
        high = (10 ** n - 1) // 6

        for x in range(low, high + 1):
            sig = signature(x)
            for k in range(2, 7):
                if signature(k * x) != sig:
                    break
            else:
                return x
        n += 1

if __name__ == "__main__":
    print(euler_52())  # 142857
